torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,696 @@
+
+
+__all__ = (
+    "RaftStereo",
+    "raft_stereo",
+    "raft_stereo_fast",
+    "Raft_Stereo_Weights",
+    "Raft_Stereo_Fast_Weights",
+)
+
+
+class BaseEncoder(raft.FeatureEncoder):
+    """Base encoder for FeatureEncoder and ContextEncoder in which weight may be shared.
+
+    See the Raft-Stereo paper section 4.6 on backbone part.
+    """
+
+    def __init__(
+        self,
+        *,
+        block: Callable[..., nn.Module] = ResidualBlock,
+        layers: Tuple[int, int, int, int] = (64, 64, 96, 128),
+        strides: Tuple[int, int, int, int] = (2, 1, 2, 2),
+        norm_layer: Callable[..., nn.Module] = nn.BatchNorm2d,
+    ):
+        # We use layers + (256,) because raft.FeatureEncoder require 5 layers
+        # but here we will set the last conv layer to identity
+        super().__init__(block=block, layers=layers + (256,), strides=strides, norm_layer=norm_layer)
+
+        # Base encoder don't have the last conv layer of feature encoder
+        self.conv = nn.Identity()
+
+        self.output_dim = layers[3]
+        num_downsampling = sum([x - 1 for x in strides])
+        self.downsampling_ratio = 2 ** (num_downsampling)
+
+
+class FeatureEncoder(nn.Module):
+    def __init__(
+        self,
+        base_encoder: BaseEncoder,
+        output_dim: int = 256,
+        shared_base: bool = False,
+        block: Callable[..., nn.Module] = ResidualBlock,
+    ):
+        super().__init__()
+        self.base_encoder = base_encoder
+        self.base_downsampling_ratio = base_encoder.downsampling_ratio
+        base_dim = base_encoder.output_dim
+
+        # If we share base encoder weight for Feature and Context Encoder
+        # we need to add residual block with InstanceNorm2d and change the kernel size for conv layer
+        # see: https://github.com/princeton-vl/RAFT-Stereo/blob/main/core/raft_stereo.py#L35-L37
+        self.shared_base = shared_base
+        self.residual_block: nn.Module = nn.Identity()
+        self.conv = nn.Conv2d(base_dim, output_dim, kernel_size=1)
+        if shared_base:
+            self.residual_block = block(base_dim, base_dim, norm_layer=nn.InstanceNorm2d, stride=1)
+            self.conv = nn.Conv2d(base_dim, output_dim, kernel_size=3, padding=1)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = self.base_encoder(x)
+        x = self.residual_block(x)
+        x = self.conv(x)
+        return x
+
+
+class MultiLevelContextEncoder(nn.Module):
+    def __init__(
+        self,
+        base_encoder: nn.Module,
+        out_with_blocks: List[bool],
+        output_dim: int = 256,
+        block: Callable[..., nn.Module] = ResidualBlock,
+    ):
+        super().__init__()
+        self.num_level = len(out_with_blocks)
+        self.base_encoder = base_encoder
+        self.base_downsampling_ratio = base_encoder.downsampling_ratio
+        base_dim = base_encoder.output_dim
+
+        self.downsample_and_out_layers = nn.ModuleList(
```

Comment:
We wrap the multi level output block and downsampler block into `ModuleList of ModuleDict` in order to make the model jit scriptable.
Previously we have 3 separate `ModuleList` and we need to iterate and access with index on one for loop, however jit script only allow to iterate module or access with literal integer.